python/seldon_core/metrics.py

```diff
@@ -1,9 +1,165 @@
+
+    def __init__(self, worker_id_func=os.getpid):
+        # We keep reference to Manager so it does not get garbage collected
+        self._manager = Manager()
+        self._lock = self._manager.Lock()
+        self.data = self._manager.dict()
+        self.worker_id_func = worker_id_func
+
+    def __del__(self):
+        self._manager.shutdown()
+
+    def update(self, custom_metrics):
+        # Read a corresponding worker's metric data with lock as Proxy objects
+        # are not thread-safe, see "Thread safety of proxies" here
+        # https://docs.python.org/3.7/library/multiprocessing.html#programming-guidelines
+        logger.debug("Updating metrics: {}".format(custom_metrics))
+        with self._lock:
+            data = self.data.get(self.worker_id_func(), {})
+        logger.debug("Read current metrics data from shared memory")
+
+        for metrics in custom_metrics:
+            key = metrics["type"], metrics["key"]
+            if metrics["type"] == "COUNTER":
+                value = data.get(key, 0)
+                data[key] = value + metrics["value"]
+            elif metrics["type"] == "TIMER":
+                vals, sumv = data.get(key, (list(np.zeros(len(BINS) - 1)), 0))
+                # Dividing by 1000 because unit is milliseconds
+                data[key] = self._update_hist(metrics["value"] / 1000, vals, sumv)
+            else:
+                data[key] = metrics["value"]
+
+        # Write worker's data with lock (again - Proxy objects are not thread-safe)
+        with self._lock:
+            self.data[self.worker_id_func()] = data
+        logger.debug("Updated metrics in the shared memory.")
+
+    def collect(self):
+        # Read all workers metrics with lock to avoid other processes / threads
+        # writing to it at the same time. Casting to `dict` works like reading of data.
+        logger.debug("SeldonMetrics.collect called")
+        with self._lock:
+            data = dict(self.data)
+        logger.debug("Read current metrics data from shared memory")
+
+        for worker, metrics in data.items():
+            labels = [
+                str(worker),
+                my_labels["seldon_deployment_name"],
+                my_labels["model_name"],
+                my_labels["image_name"],
+                my_labels["image_version"],
+                my_labels["predictor_name"],
+                my_labels["predictor_version"],
+            ]
+            for (item_type, item_name), item_value in metrics.items():
+                if item_type == "GAUGE":
+                    yield self._expose_gauge(item_name, item_value, labels)
+                elif item_type == "COUNTER":
+                    yield self._expose_counter(item_name, item_value, labels)
+                elif item_type == "TIMER":
+                    yield self._expose_histogram(item_name, item_value, labels)
+
+    def generate_metrics(self):
+        myregistry = CollectorRegistry()
+        myregistry.register(self)
```

Comment:
Does this have to be registered every time? Or can you just add this in the __init__?

Reply:
I already commented on it during the first review in https://github.com/SeldonIO/seldon-core/pull/1507#discussion_r390889127. This way I am 100% sure that this process / thread safe. I would have to carefully explore if having one shared object is
1. posible
2. thread / process safe

Reply:
Makes sense, let's skip this for now, but perhaps would be good to add a comment to explicitly mention that this was explicitly put here (so someone doesn't just move it without first checking those two pieces)

Reply:
Just to clarify: my approach to be process / thread safe was to only share between processes / threads a basic Python data types.